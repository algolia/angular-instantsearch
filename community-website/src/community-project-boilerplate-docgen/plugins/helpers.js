const highlight = require('./syntaxHighlighting.js');
const md = require('./mdRenderer.js');

// this plugin provides ATM one helper to easily compute the publicPath of assets
module.exports = function helpers(filenames, metalsmith, cb) {
  // eslint-disable-next-line no-param-reassign
  metalsmith.metadata().h = {
    markdown(src) {
      return md.render(src);
    },
    highlight(src, opts) {
      const lang = opts && opts.lang;
      return highlight(src, lang);
    },
    maybeActive(navPath, singlePathOrArrayOfPaths) {
      const pathsToTest = [].concat(singlePathOrArrayOfPaths);
      return pathsToTest.some(pathToTest => navPath.indexOf(pathToTest) === 0)
        ? 'active'
        : '';
    },
  };

  cb();
};
