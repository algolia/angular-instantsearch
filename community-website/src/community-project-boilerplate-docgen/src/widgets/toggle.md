---
title: Widgets - Toggle
layout: widget.pug
---

# Toggle

## Description

The toggle widget lets the user either:

* switch between two values for a single facetted attribute (`free_shipping / not_free_shipping`)
* toggle a faceted value on and off (only `canon` for brands)

This widget is particularly useful if you have a boolean value in the records.

## Options

| Props           | Type                            | Description
| -               | -                               | -
| `attribute`     | `string`                        | Name of the attribute for faceting (eg. “free_shipping”)
| `label`         | `string`                        | Human-readable name of the filter (eg. “Free Shipping”)
| `values`        | `{on?: boolean, off?: boolean}` | Value to filter on when unchecked. element (when using the default template). By default when switching to off, no refinement will be asked. So you will get both true and false results. If you set the off value to false then you will get only objects having false has a value for the selected attribute

## Code example

With single value:

```html
<ng-ais-toggle
  label="Free Shipping (toggle single value)"
  attribute="free_shipping"
>
</ng-ais-toggle>
```

With on & off values:

```html
<ng-ais-toggle
  label="Canon (not checked) or Sony (checked)"
  attribute="brand"
  [values]="{
    on: 'Sony',
    off: 'Canon'
  }"
>
</ng-ais-toggle>
```
