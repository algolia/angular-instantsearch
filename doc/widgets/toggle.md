## `<ng-ais-toggle />`

The toggle widget lets the user either:

* switch between two values for a single facetted attribute (`free_shipping / not_free_shipping`)
* toggle a faceted value on and off (only `canon` for brands)

This widget is particularly useful if you have a boolean value in the records.

| Props           | Type                            | Description
| -               | -                               | -
| `attribute`     | `string`                        | Name of the attribute for faceting (eg. “free_shipping”)
| `label`         | `string`                        | Human-readable name of the filter (eg. “Free Shipping”)
| `values`        | `{on?: boolean, off?: boolean}` | Value to filter on when unchecked. element (when using the default template). By default when switching to off, no refinement will be asked. So you will get both true and false results. If you set the off value to false then you will get only objects having false has a value for the selected attribute
