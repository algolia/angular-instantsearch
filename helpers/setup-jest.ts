import "jest-preset-angular";

// stupid workaround for TS
import * as helper from "algoliasearch-helper";
require("algoliasearch-helper").default = helper;

const mock = () => {
  let storage = {};
  return {
    getItem(key) {
      return storage[key] || null;
    },
    setItem(key, value) {
      storage[key] = value;
    },
    removeItem(key) {
      delete storage[key];
    },
    clear() {
      storage = {};
    }
  };
};

Object.defineProperty(window, "localStorage", { value: mock() });
Object.defineProperty(window, "sessionStorage", { value: mock() });
Object.defineProperty(window, "getComputedStyle", {
  value: () => ["-webkit-appearance"]
});
