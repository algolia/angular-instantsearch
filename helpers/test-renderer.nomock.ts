import { Component, ViewChild } from '@angular/core';
import { TestBed } from '@angular/core/testing';

import { NgAisInstantSearchModule } from '../src/instantsearch/instantsearch.module';

// same as test-renderer, but without mocking InstantSearch
// In the future, we want to migrate all tests to this

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations,
  methods = {},
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
  methods?: {};
}) {
  return function(state?: {}, isFirstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined,
        methods,
      },
      isFirstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state,
    methods,
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
    methods?: {};
  },
  isFirstRender = false
) {
  @Component({
    template: `
      <ais-instantsearch [config]="config">
        ${template}
      </ais-instantsearch>
    `,
  })
  class TestContainer {
    config = {
      searchClient: {
        search(requests) {
          return Promise.resolve({
            results: requests.map(() => ({ hits: [] })),
          });
        },
      },
      indexName: 'instant_search',
    };
    @ViewChild(TestedWidget) testedWidget;
    constructor() {
      Object.keys(methods).forEach(methodName => {
        this[methodName] = methods[methodName];
      });
    }
  }

  TestBed.configureCompiler({
    preserveWhitespaces: false,
  } as any).configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || []),
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])],
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, isFirstRender);
    fixture.detectChanges();
  }

  return fixture;
}
