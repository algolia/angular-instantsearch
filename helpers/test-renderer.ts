import { Component, ViewChild } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../src/instantsearch/instantsearch.module";

// mock component, dont create a real instantearch instance
jest.mock("../src/base-widget");
jest.mock("../src/instantsearch/instantsearch");

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
}) {
  return function(state?: {}, firstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined
      },
      firstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
  },
  firstRender = false
) {
  @Component({
    template: `
      <ng-ais-instantsearch>
        ${template}
      </ng-ais-instantsearch>
    `
  })
  class TestContainer {
    @ViewChild(TestedWidget) testedWidget;
  }

  TestBed.configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || [])
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])]
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, firstRender);
    fixture.detectChanges();
  }

  return fixture;
}
