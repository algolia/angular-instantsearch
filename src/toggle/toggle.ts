import { Component, Input, Inject, PLATFORM_ID } from "@angular/core";
import { connectToggle } from "instantsearch.js/es/connectors";
import { noop } from "lodash-es";

import { BaseWidget } from "../base-widget";
import { NgAisInstance } from "../instantsearch/instantsearch-instance";

export type ToggleState = {
  createURL: Function;
  refine: Function;
  value: {
    name?: string;
    count?: number;
    isRefined?: boolean;
  };
};

@Component({
  selector: "ng-ais-toggle",
  template: `
    <div [class]="cx()">
      <ul [class]="cx('list')">
        <li
          [class]="cx('item')"
          (click)="handleClick($event)">
          <label [class]="cx('label')">
            <input
              [class]="cx('checkbox')"
              type="checkbox"
              value="{{state.value.name}}"
              [checked]="state.value.isRefined"
            />

            <span [class]="cx('labelText')">
              {{label || state.value.name}}
            </span>

            <span [class]="cx('count')">{{state.value.count}}</span>
          </label>
        </li>
      </ul>
    </div>
  `
})
export class NgAisToggle extends BaseWidget {
  // connector options
  @Input() public attribute: string;
  @Input() public label: string;
  @Input()
  public values: { on?: boolean; off?: boolean } = { on: true, off: undefined };

  public state: ToggleState = {
    createURL: noop,
    refine: noop,
    value: {}
  };

  constructor(
    @Inject(PLATFORM_ID) public platformId: Object,
    searchInstance: NgAisInstance
  ) {
    super(searchInstance, "ToggleRefinement");
  }

  public ngOnInit() {
    this.createWidget(connectToggle, {
      attributeName: this.attribute,
      label: this.label,
      values: this.values
    });
    super.ngOnInit();
  }

  public handleClick(event: MouseEvent) {
    event.preventDefault();
    event.stopPropagation();
    this.state.refine(this.state.value);
  }
}
