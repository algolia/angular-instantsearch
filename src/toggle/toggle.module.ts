import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";

import { NgAisToggle } from "./toggle";

@NgModule({
  declarations: [NgAisToggle],
  entryComponents: [NgAisToggle],
  exports: [NgAisToggle],
  imports: [CommonModule]
})
export class NgAisToggleModule {}
