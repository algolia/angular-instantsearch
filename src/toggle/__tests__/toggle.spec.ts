import { createRenderer } from "../../../helpers/test-renderer";
import { NgAisToggle } from "../toggle";

const defaultState = {
  createURL: jest.fn(),
  refine: jest.fn(),
  value: {
    isRefined: true,
    name: "foobar",
    count: 666
  }
};

const render = createRenderer({
  defaultState,
  template: "<ais-toggle></ais-toggle>",
  TestedWidget: NgAisToggle
});

describe("Toggle", () => {
  it("should render without state", () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it("should render with state", () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it("should render with a label", () => {
    const fixture = render({});
    fixture.componentInstance.testedWidget.label = "Foo Label";
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it("should call refine on click", () => {
    const refine = jest.fn();
    const fixture = render({ refine });

    const toggle = fixture.debugElement.nativeElement.querySelector("li");
    toggle.click();

    expect(refine).toHaveBeenCalled();
    expect(refine).toHaveBeenCalledWith(defaultState.value);
  });
});
