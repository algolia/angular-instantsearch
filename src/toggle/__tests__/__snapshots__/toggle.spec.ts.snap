// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle should render with a label 1`] = `
<ng-component
  testedWidget={[Function NgAisToggle]}
>
  <ais-instantsearch>
    <ais-toggle>
      <div
        class="ais-ToggleRefinement"
      >
        <ul
          class="ais-ToggleRefinement-list"
        >
          <li
            class="ais-ToggleRefinement-item"
          >
            <label
              class="ais-ToggleRefinement-label"
            >
              <input
                class="ais-ToggleRefinement-checkbox"
                type="checkbox"
                value="foobar"
              />
              <span
                class="ais-ToggleRefinement-labelText"
              >
                 Foo Label 
              </span>
              <span
                class="ais-ToggleRefinement-count"
              >
                666
              </span>
            </label>
          </li>
        </ul>
      </div>
    </ais-toggle>
  </ais-instantsearch>
</ng-component>
`;

exports[`Toggle should render with state 1`] = `
<ng-component
  testedWidget={[Function NgAisToggle]}
>
  <ais-instantsearch>
    <ais-toggle>
      <div
        class="ais-ToggleRefinement"
      >
        <ul
          class="ais-ToggleRefinement-list"
        >
          <li
            class="ais-ToggleRefinement-item"
          >
            <label
              class="ais-ToggleRefinement-label"
            >
              <input
                class="ais-ToggleRefinement-checkbox"
                type="checkbox"
                value="foobar"
              />
              <span
                class="ais-ToggleRefinement-labelText"
              >
                 foobar 
              </span>
              <span
                class="ais-ToggleRefinement-count"
              >
                666
              </span>
            </label>
          </li>
        </ul>
      </div>
    </ais-toggle>
  </ais-instantsearch>
</ng-component>
`;

exports[`Toggle should render without state 1`] = `
<ng-component
  testedWidget={[Function NgAisToggle]}
>
  <ais-instantsearch>
    <ais-toggle>
      <div
        class="ais-ToggleRefinement"
      >
        <ul
          class="ais-ToggleRefinement-list"
        >
          <li
            class="ais-ToggleRefinement-item"
          >
            <label
              class="ais-ToggleRefinement-label"
            >
              <input
                class="ais-ToggleRefinement-checkbox"
                type="checkbox"
                value=""
              />
              <span
                class="ais-ToggleRefinement-labelText"
              >
                  
              </span>
              <span
                class="ais-ToggleRefinement-count"
              >
                
              </span>
            </label>
          </li>
        </ul>
      </div>
    </ais-toggle>
  </ais-instantsearch>
</ng-component>
`;
